kafka/kafka.go

```diff
@@ -247,85 +244,68 @@ func (cm *consumerMessage) DiscardPayload() {
+func (_ *consumerGroupHandler) Setup(_ sarama.ConsumerGroupSession) error { return nil }
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+// but before the offsets are committed for the very last time.
+func (c *consumerGroupHandler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+// Once the Messages() channel is closed, the Handler must finish its processing
+// loop and exit.
+func (c *consumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	// This function can be called concurrently for multiple claims, so the code
+	// below, absent locking etc may seem wrong, but it's actually fine.
+	// Different partition claims can be processed concurrently, but we funnel
+	// them all into c.messages.  The caller puts all acks into c.acks and it
+	// doesn't matter which one of us processes the offset marking because they
+	// all work with the same session.
 	for {
 		select {
-		case <-ctx.Done():
-			return nil
-		case msg := <-fromKafka:
-			message := &consumerMessage{cm: msg}
-
-			select {
-			case <-ctx.Done():
+		case <-c.ctx.Done():
+			return c.ctx.Err()
+		case m, ok := <-claim.Messages():
+			if !ok {
 				return nil
-			case toAck <- message:
 			}
 			select {
-			case <-ctx.Done():
-				return nil
-			case messages <- message:
+			case c.messages <- &consumerMessage{cm: m}:
+			case <-c.ctx.Done():
+				return c.ctx.Err()
+			}
+		case a := <-c.acks:
+			cm := a.(*consumerMessage)
+			if cm.cm != nil {
+				sess.MarkMessage(cm.cm, "")
+			} else {
+				off := cm.offset
+				sess.MarkOffset(off.topic, off.partition, off.offset, "")
 			}
 		}
 	}
 }
 
-func (ams *asyncMessageSource) passAcksToKafka(ctx context.Context, c *cluster.Consumer, acks <-chan substrate.Message, toAck <-chan *consumerMessage) error {
-	var forAcking []*consumerMessage
-
-	for {
-		select {
-		case msg := <-toAck:
-			forAcking = append(forAcking, msg)
-		case ack := <-acks:
-			switch {
-			case len(forAcking) == 0:
-				return substrate.InvalidAckError{
-					Acked:    ack,
-					Expected: nil,
-				}
-			case ack != forAcking[0]:
-				return substrate.InvalidAckError{
-					Acked:    ack,
-					Expected: forAcking[0],
-				}
-			default:
-				if forAcking[0].offset != nil {
-					off := forAcking[0].offset
-					c.MarkPartitionOffset(off.topic, off.partition, off.offset, "")
-				} else {
-					c.MarkOffset(forAcking[0].cm, "")
-				}
-				forAcking = forAcking[1:]
-			}
-		case err := <-c.Errors():
-			return err
-		case <-ctx.Done():
-			return c.Close()
-		}
+func (ams *asyncMessageSource) ConsumeMessages(ctx context.Context, messages chan<- substrate.Message, acks <-chan substrate.Message) error {
+	c, err := sarama.NewConsumerGroupFromClient(ams.consumerGroup, ams.client)
+	if err != nil {
+		return err
 	}
+
+	defer func() {
+		_ = c.Close()
```

Comment:
I think this might cause issues as it can be called after the client is closed, so I think they should both be closed in the close function of the message source.